pbj-core/pbj-grpc-helidon/src/main/java/com/hedera/pbj/grpc/helidon/PbjProtocolHandler.java

```diff
@@ -0,0 +1,590 @@
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.pbj.grpc.helidon.config.PbjConfig;
+import com.hedera.pbj.runtime.grpc.GrpcException;
+import com.hedera.pbj.runtime.grpc.GrpcStatus;
+import com.hedera.pbj.runtime.grpc.ServiceInterface;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import io.helidon.common.buffers.BufferData;
+import io.helidon.common.uri.UriEncoding;
+import io.helidon.http.Header;
+import io.helidon.http.HeaderNames;
+import io.helidon.http.HeaderValues;
+import io.helidon.http.HttpException;
+import io.helidon.http.HttpMediaType;
+import io.helidon.http.HttpMediaTypes;
+import io.helidon.http.Status;
+import io.helidon.http.WritableHeaders;
+import io.helidon.http.http2.Http2Flag;
+import io.helidon.http.http2.Http2FrameData;
+import io.helidon.http.http2.Http2FrameHeader;
+import io.helidon.http.http2.Http2FrameTypes;
+import io.helidon.http.http2.Http2Headers;
+import io.helidon.http.http2.Http2RstStream;
+import io.helidon.http.http2.Http2StreamState;
+import io.helidon.http.http2.Http2StreamWriter;
+import io.helidon.http.http2.Http2WindowUpdate;
+import io.helidon.http.http2.StreamFlowControl;
+import io.helidon.webserver.http2.spi.Http2SubProtocolSelector;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.Flow;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+/**
+ * Implementation of gRPC based on PBJ. This class specifically contains the glue logic for bridging between
+ * Helidon and the generated PBJ service handler endpoints. An instance of this class is created for each new
+ * connection, and each connection is made to a specific method endpoint.
+ */
+final class PbjProtocolHandler implements Http2SubProtocolSelector.SubProtocolHandler {
+    private static final System.Logger LOGGER = System.getLogger(PbjProtocolHandler.class.getName());
+    /** The only grpc-encoding supported by this implementation. */
+    private static final String IDENTITY = "identity";
+    /** A pre-created and cached *response* header for "grpc-encoding: identity". */
+    private static final Header GRPC_ENCODING_IDENTITY = HeaderValues.createCached("grpc-encoding", IDENTITY);
+    /** The regular expression used to parse the grpc-timeout header. */
+    private static final String GRPC_TIMEOUT_REGEX = "(\\d{1,8})([HMSmun])";
+    private static final Pattern GRPC_TIMEOUT_PATTERN = Pattern.compile(GRPC_TIMEOUT_REGEX);
+
+    // Helidon-specific fields related to the connection itself
+    private final Http2Headers headers;
+    private final Http2StreamWriter streamWriter;
+    private final int streamId;
+    private final StreamFlowControl flowControl;
+    private Http2StreamState currentStreamState;
```

Comment:
I changed this to be an atomic reference, and removed the `synchronization` keyword from the `close()` method, since the only other thing in `close` apart from setting `currentStreamState` is threadsafe